android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/util/platforms/android/AndroidShellExec.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.util.platforms.android;
+
+import android.os.SystemClock;
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.util.platforms.Platform;
+
+import java.io.IOException;
+import java.lang.Process;
+import java.lang.reflect.Field;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+// FIXME: android: to be implemented
+@SuppressWarnings("PMD.AvoidCatchingThrowable")
+public class AndroidShellExec extends AndroidGenericExec {
+    private static final Logger staticLogger = LogManager.getLogger(AndroidShellExec.class);
+    private int pid;
+
+    private static void ensurePresent(String... fns) {
+        for (String fn : fns) {
+            Path ulb = Paths.get(fn);
+            if (Files.isDirectory(ulb) && !paths.contains(ulb)) {
+                paths.add(ulb);
+            }
+        }
+    }
+
+    @Nullable
+    @Override
+    public Path which(String fn) {
+        fn = deTilde(fn);
+        if (fn.startsWith("/")) {
+            Path f = Paths.get(fn);
+            return Files.isExecutable(f) ? f : null;
+        }
+        for (Path d : paths) {
+            Path f = d.resolve(fn);
+            if (Files.isExecutable(f)) {
+                return f;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    protected Process createProcess() throws IOException {
+        String[] command = getCommand();
+        logger.atTrace().kv("decorated command", String.join(" ", command)).log();
+        ProcessBuilder pb = new ProcessBuilder();
+        pb.environment().putAll(environment);
+        process = pb.directory(dir).command(command).start();
+        pid = -1;
+        try {
+            Field f = process.getClass().getDeclaredField("pid");
+            f.setAccessible(true);
+            pid = (int) f.getLong(process);
+            f.setAccessible(false);
+        } catch (NoSuchFieldException | IllegalAccessException cause) {
+            logger.atError().setCause(cause).log("Failed to get process pid");
+        }
+        return process;
+    }
+
+    @Override
+    public int getPid() {
+        return pid;
+    }
+
+    @Override
+    public String[] getCommand() {
+        String[] decorated = cmds;
+        if (shellDecorator != null) {
+            decorated = shellDecorator.decorate(decorated);
+        }
+        if (userDecorator != null) {
+            decorated = userDecorator.decorate(decorated);
+        }
+        return decorated;
+    }
+
+    @Override
+    public synchronized void close() throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.